x/mongo/driver/topology/server_test.go

```diff
@@ -984,6 +836,391 @@ func TestServer(t *testing.T) {
 	})
 }
 
+func TestServer_ProcessError(t *testing.T) {
+	processID := primitive.NewObjectID()
+	newProcessID := primitive.NewObjectID()
+
+	// newServerDescription is a convenience function for creating a server description with a
+	// specified kind, topology version process ID and counter, and last error.
+	newServerDescription := func(
+		kind description.ServerKind,
+		processID primitive.ObjectID,
+		counter int64,
+		lastError error,
+	) description.Server {
+		return description.Server{
+			Kind: kind,
+			TopologyVersion: &description.TopologyVersion{
+				ProcessID: processID,
+				Counter:   counter,
+			},
+			LastError: lastError,
+		}
+	}
+
+	testCases := []struct {
+		name string
+
+		startDescription description.Server // Initial server description at the start of the test.
+
+		inputErr  error             // ProcessError error input.
+		inputConn driver.Connection // ProcessError conn input.
+
+		want            driver.ProcessErrorResult // Expected ProcessError return value.
+		wantGeneration  uint64                    // Expected resulting connection pool generation.
+		wantDescription description.Server        // Expected resulting server description.
+	}{
+		// Test that a nil error does not change the Server state.
+		{
+			name: "nil error",
+			startDescription: description.Server{
+				Kind: description.RSPrimary,
+			},
+			inputErr:       nil,
+			want:           driver.NoChange,
+			wantGeneration: 0,
+			wantDescription: description.Server{
+				Kind: description.RSPrimary,
+			},
+		},
+		// Test that errors that occur on stale connections are ignored.
+		{
+			name: "stale connection",
+			startDescription: description.Server{
+				Kind: description.RSPrimary,
+			},
+			inputErr: errors.New("foo"),
+			inputConn: newProcessErrorTestConn(
+				&description.VersionRange{
+					Max: 17,
+				},
+				true),
+			want:           driver.NoChange,
+			wantGeneration: 0,
+			wantDescription: description.Server{
+				Kind: description.RSPrimary,
+			},
+		},
+		// Test that errors that do not indicate a database state change or connection error are
+		// ignored.
+		{
+			name: "non state change error",
+			startDescription: description.Server{
+				Kind: description.RSPrimary,
+			},
+			inputErr: driver.Error{
+				Code: 1,
+			},
+			inputConn:      newProcessErrorTestConn(&description.VersionRange{Max: 17}, false),
+			want:           driver.NoChange,
+			wantGeneration: 0,
+			wantDescription: description.Server{
+				Kind: description.RSPrimary,
+			},
+		},
+		// Test that a "not primary" error with an old topology version is ignored.
+		{
+			name:             "stale not primary error",
+			startDescription: newServerDescription(description.RSPrimary, processID, 1, nil),
+			inputErr: driver.Error{
+				Code: 10107, // NotPrimary
```

Comment:
Optional: 10107 is "NotWritablePrimary", see [here](https://github.com/mongodb/mongo/blob/master/src/mongo/base/error_codes.yml#LL543C26-L543C44).

Reply:
Good catch. That error is documented as "NotPrimary" in many places in the Go driver code. I'll correct it in these tests and then open a subsequent PR to correct it elsewhere in the Go driver.